File: packages/api/src/api/schemas/portfolio.py
"""Portfolio summary and timeseries response shapes."""
from __future__ import annotations

from datetime import date
from decimal import Decimal
from uuid import UUID

from api.schemas.common import APIBase
from api.schemas.property import PropertyType


class ValueByType(APIBase):
    type: PropertyType
    value: Decimal
    count: int


class ValueByEntity(APIBase):
    entity_id: UUID
    name: str
    value: Decimal
    count: int


class TopProperty(APIBase):
    property_id: UUID
    name: str
    value: Decimal


class PortfolioSummary(APIBase):
    total_market_value: Decimal
    property_count: int
    entity_count: int
    last_snapshot_date: date | None
    value_by_type: list[ValueByType]
    value_by_entity: list[ValueByEntity]
    top_properties: list[TopProperty]


class TimeseriesPoint(APIBase):
    bucket_date: date
    total_market_value: Decimal
    property_count: int


class PortfolioTimeseries(APIBase):
    bucket: str
    points: list[TimeseriesPoint]

File: packages/web/src/schemas/portfolio.ts
import { z } from 'zod'
import { PropertyTypeSchema } from './property'

// Mirrors api/schemas/portfolio.py
// Reference guessed by_type/by_entity — real field names are value_by_type/value_by_entity
// Decimal fields (value) serialised as strings on the wire

export const ValueByTypeSchema = z.object({
  type: PropertyTypeSchema,
  value: z.string(),
  count: z.number().int(),
})
export type ValueByType = z.infer<typeof ValueByTypeSchema>

export const ValueByEntitySchema = z.object({
  entity_id: z.string(),
  name: z.string(),
  value: z.string(),
  count: z.number().int(),
})
export type ValueByEntity = z.infer<typeof ValueByEntitySchema>

export const TopPropertySchema = z.object({
  property_id: z.string(),
  name: z.string(),
  value: z.string(),
})
export type TopProperty = z.infer<typeof TopPropertySchema>

export const PortfolioSummarySchema = z.object({
  total_market_value: z.string(),
  property_count: z.number().int(),
  entity_count: z.number().int(),
  last_snapshot_date: z.string().nullable(),
  value_by_type: z.array(ValueByTypeSchema),
  value_by_entity: z.array(ValueByEntitySchema),
  top_properties: z.array(TopPropertySchema),
})
export type PortfolioSummary = z.infer<typeof PortfolioSummarySchema>

export const TimeseriesPointSchema = z.object({
  bucket_date: z.string(),
  total_market_value: z.string(),
  property_count: z.number().int(),
})
export type TimeseriesPoint = z.infer<typeof TimeseriesPointSchema>

export const PortfolioTimeseriesSchema = z.object({
  bucket: z.string(),
  points: z.array(TimeseriesPointSchema),
})
export type PortfolioTimeseries = z.infer<typeof PortfolioTimeseriesSchema>
